base/bundle.cpp

```diff
@@ -0,0 +1,113 @@
+#pragma once
+
+#include "base/bundle.hpp"
+
+#include <functional>
+#include <queue>
+
+#include "base/map_util.hpp"
+#include "base/status.hpp"
+
+namespace principia {
+namespace base {
+
+Bundle::Bundle(int threads) {
+  {
+    std::unique_lock<std::mutex> l(lock_);
+    joining_ = false;
+  }
+  for (int i = 0; i < threads; ++i) {
+    threads_.emplace_back([this]() {
+      std::shared_ptr<Task> current_task;
+      for (;;) {
+        {
+          std::unique_lock<std::mutex> l(lock_);
+          if (joining_ && tasks_.empty()) {
+            return;
+          }
+          tasks_not_empty_or_joining_.wait(
+              l,
+              [this]() { return !tasks_.empty() || joining_; });
+          if (tasks_.empty()) {
+            return;
+          }
+          current_task = std::move(tasks_.front());
+          tasks_.pop_front();
+          current_task->in_queue_ = std::experimental::nullopt;
```

Comment:
It seems to me that what you are doing with `in_queue_` and `abort_requested_` would be better implemented with a state machine: *queued*->*started*->*completed*->*done* with some additional transitions to *aborted* and *joining*.  That would bring clarity to the code.